definitions/npm/styled-components_v2.x.x/flow_v0.53.x-/test_styled-components_v2.x.x.js

```diff
@@ -0,0 +1,260 @@
+import styled, {
+  ThemeProvider,
+  withTheme,
+  keyframes,
+  ServerStyleSheet,
+  StyleSheetManager
+} from 'styled-components'
+import React from 'react'
+import type {
+  Theme,
+  Interpolation,
+
+  // Temporary
+  ReactComponentFunctional,
+  ReactComponentClass,
+  ReactComponentStyled,
+  ReactComponentStyledTaggedTemplateLiteral,
+  ReactComponentUnion,
+  ReactComponentIntersection,
+} from 'styled-components'
+
+const TitleTaggedTemplateLiteral: ReactComponentStyledTaggedTemplateLiteral<{}> = styled.h1;
+
+const TitleStyled: ReactComponentStyled<*> = styled.h1`
+  font-size: 1.5em;
+`;
+
+const TitleGeneric: ReactComponentIntersection<*> = styled.h1`
+  font-size: 1.5em;
+`;
+
+const TitleFunctional: ReactComponentFunctional<*> = styled.h1`
+  font-size: 1.5em;
+`;
+
+const TitleClass: ReactComponentClass<*> = styled.h1`
+  font-size: 1.5em;
+`;
+
+declare var needsReactComponentFunctional: ReactComponentFunctional<*> => void
+declare var needsReactComponentClass: ReactComponentClass<*> => void
+
+needsReactComponentFunctional(TitleStyled)
+needsReactComponentClass(TitleStyled)
+
+const ExtendedTitle: ReactComponentIntersection<*> = styled(TitleStyled)`
+  font-size: 2em;
+`;
+
+const Wrapper: ReactComponentIntersection<*> = styled.section`
+  padding: 4em;
+  background: ${({theme}) => theme.background};
+`;
+
+// ---- EXTEND ----
+const Attrs0ReactComponent: ReactComponentStyled<*> = styled.div.extend``;
+const Attrs0ExtendReactComponent: ReactComponentIntersection<*> = Attrs0ReactComponent.extend``;
+
+const Attrs0SyledComponent: ReactComponentStyledTaggedTemplateLiteral<*> = styled.div;
+const Attrs0ExtendStyledComponent: ReactComponentIntersection<*> = Attrs0SyledComponent.extend``;
+
+// ---- ATTRIBUTES ----
+const Attrs1: ReactComponentStyledTaggedTemplateLiteral<*> = styled.section.attrs({
+  testProp: 'foo'
+});
+
+// $ExpectError
+const Attrs1Error: ReactComponentStyledTaggedTemplateLiteral<*> = styled.section.attrs({
+  testProp: 'foo'
+})``;
+
+declare var needsString: string => void
+
+needsReactComponentFunctional(styled.section.attrs({})``)
+needsReactComponentClass(styled.section.attrs({})``)
+
+// $ExpectError
+needsString(styled.section.attrs({})``)
+
+const Attrs2: ReactComponentStyledTaggedTemplateLiteral<*> = styled.section
+  .attrs({
+    testProp1: 'foo'
+  })
+  .attrs({
+    testProp2: 'bar'
+  });
+
+const Attrs3Styled: ReactComponentStyled<*> = styled.section.attrs({
+  testProp: 'foo'
+})`
+  background-color: red;
+`;
+
+const Attrs3Generic: ReactComponentIntersection<*> = styled.section.attrs({
+  testProp: 'foo'
+})`
+  background-color: red;
+`;
+
+const Attrs3Functional: ReactComponentFunctional<*> = styled.section.attrs({
+  testProp: 'foo'
+})`
+  background-color: red;
+`;
+
+const Attrs3Class: ReactComponentClass<*> = styled.section.attrs({
+  testProp: 'foo'
+})`
+  background-color: red;
+`;
+
+const theme: Theme = {
+  background: "papayawhip"
+};
+
+// ---- WithComponent ----
```

Comment:
Should not there be a test for `styled(Component).withComponent('a')` for example?  I was trying this PR in our codebase and it failed on `withComponent()` calls when there is a custom component wrapped in `styled`

Reply:
Your definition seems to throw errors on my end. Could you also post the test files for it?

Starting to think rewriting it would work be easier.

Reply:
@michalkvasnicak What is the type of your custom component? Can you shoe the test case that fails?

Reply:
@michalkvasnicak I was just about to write a test case for this, but I don't think this should work for custom components...
```js
class CustomComponent extends React.Component<{ foo: string }> {
  render() {
    return <div>{this.props.foo}</div>
  }
}
const CustomComponentNewTag = styled(CustomComponent).withComponent('a') // ???
```
How exactly should ``CustomComponentNewTag`` behave in this case?

I understand how it would work when using a default component from ``$npm$styledComponents$StyledComponentsComponentList``, but not for custom components.

https://www.styled-components.com/docs/basics#extending-styles

Reply:
Right now ``styled(CustomComponent)`` has the ``withComponent`` property and shouldn't...

Reply:
I think if we remove custom components from the input list in `$npm$styledComponents$ReactComponentStyledStaticPropsWithComponent` we should get what we want.

Reply:
@NullDivision https://github.com/flowtype/flow-typed/pull/1194#issuecomment-327029569